packages/block-editor/src/utils/dom.js

```diff
@@ -57,3 +57,132 @@ export function getBlockClientId( node ) {
+
+	if ( rectUnionCache.has( cacheKey ) ) {
+		return rectUnionCache.get( cacheKey );
+	}
+
+	let left = Math.min( rect1.left, rect2.left );
+	let right = Math.max( rect1.right, rect2.right );
+	const bottom = Math.max( rect1.bottom, rect2.bottom );
+	const top = Math.min( rect1.top, rect2.top );
+
+	/*
+	 * To calculate visible bounds using rectUnion, take into account the outer
+	 * horizontal limits of the container in which an element is supposed to be "visible".
+	 * For example, if an element is positioned -10px to the left of the window x value (0),
+	 * this function discounts the negative overhang because it's not visible and
+	 * therefore not to be counted in the visibility calculations.
+	 * Top and bottom values are not accounted for to accommodate vertical scroll.
+	 */
+	if ( containerRect ) {
+		left = Math.max( left, containerRect.left );
+		right = Math.min( right, containerRect.right );
+	}
+
+	const result = new window.DOMRect( left, top, right - left, bottom - top );
+
+	rectUnionCache.set( cacheKey, result );
+
+	return result;
+}
+
+/**
+ * Returns whether an element is visible.
+ *
+ * @param {Element} element Element.
+ * @return {boolean} Whether the element is visible.
+ */
+function isElementVisible( element ) {
+	const viewport = element.ownerDocument.defaultView;
+	if ( ! viewport ) {
+		return false;
+	}
+
+	// Check for <VisuallyHidden> component.
+	if ( element.classList.contains( 'components-visually-hidden' ) ) {
+		return false;
+	}
+
+	const bounds = element.getBoundingClientRect();
+	if ( bounds.width === 0 || bounds.height === 0 ) {
+		return false;
+	}
+
+	const style = viewport.getComputedStyle( element );
+	if (
+		style.display === 'none' ||
+		style.visibility === 'hidden' ||
+		style.opacity === '0'
+	) {
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Returns the rect of the element including all visible nested elements.
+ *
+ * Visible nested elements, including elements that overflow the parent, are
+ * taken into account.
+ *
+ * This function is useful for calculating the visible area of a block that
+ * contains nested elements that overflow the block, e.g. the Navigation block,
+ * which can contain overflowing Submenu blocks.
+ *
+ * The returned rect represents the full extent of the element and its visible
+ * children, which may extend beyond the viewport.
+ *
+ * @param {Element} element Element.
+ * @return {DOMRect} Bounding client rect of the element and its visible children.
+ */
+export function getVisibleElementBounds( element ) {
+	const viewport = element.ownerDocument.defaultView;
+	if ( ! viewport ) {
+		return new window.DOMRect();
+	}
+
+	let bounds = element.getBoundingClientRect();
+	const viewportRect = new window.DOMRectReadOnly(
+		0,
+		0,
+		viewport.innerWidth,
+		viewport.innerHeight
+	);
+
+	const stack = [ element ];
+	let currentElement;
+
+	while ( ( currentElement = stack.pop() ) ) {
+		for ( const child of currentElement.children ) {
```

Comment:
This means we have to iterate through every child element which could be a bit costly? I don't know if we can cut off some operations here 🤔.

An immediate thought would be asking the reference element (nav link) to declaratively specify the overflow element (submenu). We could potentially use a `data-popover-target` attribute to find the overflow elements and only perform this extra check if it's specified. I think this would also prevent any unexpected side effects and maintain backward compatibility. The downside is that sometimes it's not better to declare an extra attribute.

Another rough idea is to hook into the `floating-ui`'s [middleware](https://floating-ui.com/docs/middleware) system. Instead of traversing through every descendant, We only check for those that overflow by leveraging [`document.elementsFromPoint`](https://developer.mozilla.org/en-US/docs/Web/API/Document/elementsFromPoint). An untested pseudo code looks something like:

```ts
const overflowDescendants: Middleware = {
  name: 'overflowDescendants',
  fn({
    x,
    y,
    elements: { reference },
    rects: { reference: refRect, floating: floatingRect },
  }) {
    // Get the four vertices of the floating element (block toolbar).
    const vertices = [
      [x, y],
      [x + floatingRect.width, y],
      [x, y + floatingRect.height],
      [x + floatingRect.width, y + floatingRect.height],
    ] as const;
    // Get all the elements at those points.
    const elements = vertices.flatMap((vertice) =>
      document.elementsFromPoint(...vertice)
    );
    // Filter out those who are not the descendants of the reference element (anchor block).
    const descendents = elements.filter(
      (element) =>
        reference !== element && (reference as HTMLElement).contains(element)
    );
    // Bail out if no elements are found.
    if (!descendents.length) {
      return {};
    }

    let minX = refRect.x;
    let maxX = refRect.x + refRect.width;
    let minY = refRect.y;
    let maxY = refRect.y + refRect.height;

    // Calculate the new union bounding rect.
    descendents.forEach((descendent) => {
      const rect = descendent.getBoundingClientRect();
      minX = Math.min(minX, rect.x);
      maxX = Math.max(maxX, rect.right);
      minY = Math.min(minY, rect.y);
      maxY = Math.max(maxY, rect.bottom);
    });

    return {
      reset: {
        rects: {
          floating: floatingRect,
          reference: {
            x: minX,
            y: minY,
            width: maxX - minX,
            height: maxY - minY,
          },
        },
      },
    };
  },
};
```

I'm not even sure if this is more performant, though 😅. We probably need some testing if this is even worth trying. This can be done in a separate explorative PR if needed too.

Reply:
Thanks for looking into it! Looks like a follow up to me 😄